src/event_handler.rs

```diff
@@ -402,12 +501,66 @@ pub async fn handle_event_logs(
 }
 
-// Retry a transaction if nonce issues occur
-fn should_retry(error: &anyhow::Error) -> bool {
-    let error_string = format!("{:?}", error);
-    if error_string.contains("code: -32000") && error_string.contains("nonce") {
-        return true;
+// Function to regularly check a transaction for block confirmation and resend it if not included
+async fn resend_pending_transaction(
+    app_state: Data<AppState>,
+    txn: FunctionCall<Arc<HttpSignerProvider>, HttpSignerProvider, ()>,
+    mut pending_txn_hash: H256,
+    nonce: U256,
+    mut gas_limit: U256,
+    mut gas_price: U256,
+    resend_deadline: u128,
+) {
+    // Calculating resend retries number based on the overall deadline and interval in which to resend pending/dropped txns
+    let mut resend_retries = resend_deadline / (RESEND_TXN_INTERVAL as u128);
+    while resend_retries > 0 {
+        sleep(Duration::from_secs(RESEND_TXN_INTERVAL)).await;
+
+        // Get the transaction receipt for the pending transaction to check if it's still pending or been dropped
+        let http_rpc_client = app_state.http_rpc_client.lock().unwrap().clone().unwrap();
+        let Ok(txn_receipt) = http_rpc_client
+            .get_transaction_receipt(pending_txn_hash)
+            .await
+        else {
+            resend_retries -= 1;
+            continue;
+        };
+
+        if txn_receipt.is_some() {
+            break;
+        }
+
+        // Update gas limit and price for sending the replacement transaction with the original nonce
+        gas_limit += GAS_LIMIT_BUFFER.into();
+        gas_price += (U256::from(RESEND_GAS_PRICE_INCREMENT_PERCENT) * gas_price) / 100;
+
+        // Update replacement transaction metadata and send it to the common chain, stop resending if failed to do so once
+        let replacement_txn = txn.clone();
+        let replacement_txn = replacement_txn
+            .nonce(nonce)
+            .gas(gas_limit)
+            .gas_price(gas_price);
+        let Ok(pending_txn) = replacement_txn.send().await else {
+            resend_retries = 0;
+            break;
+        };
+
+        pending_txn_hash = pending_txn.tx_hash();
+        resend_retries -= 1;
     }
 
-    false
+    // If the original transaction fails to get included in the common chain, send a dummy transaction that is confirmed to be included and will replace the original nonce transaction
+    if resend_retries == 0 {
+        // Increasing gas price so that the current transaction can replace the original in the mempool
+        gas_price += (U256::from(RESEND_GAS_PRICE_INCREMENT_PERCENT) * gas_price) / 100;
+        let http_rpc_client = app_state.http_rpc_client.lock().unwrap().clone().unwrap();
+
+        // Send 0 ETH to self from the gas account as a confirmed replacement transaction
+        let dummy_replacement_txn = TransactionRequest::pay(http_rpc_client.address(), 0u64)
+            .nonce(nonce)
+            .gas_price(gas_price);
+        let _ = http_rpc_client
```

Comment:
at least check for any http errors, otherwise whole pipeline will be blocked. This whole block seems very risky. Something goes wrong (i.e. dummy txn doesn't go through), and whole thing stuck.

Reply:
Will add retries here also for connection errors. Dummy txn if sent successfully is our best bet to replace a stuck or dropped transaction. Otherwise we'll just get stuck in a indefinite loop of retrying transactions. Do you have a suggestion to avoid these loops of retries or any other way to handle pending/dropped transactions?

Reply:
I can see it  this way, if you are not able to send this txn because connection issue, you won't be able to send subsequent transactions too.